### ISSUE TITLE
[Bug] Most pictures are not rendered

### ISSUE THREAD

**Issue by prosch88** [leonlatsch/Photok#330]
I recently updated to 1.7. 
All thumbnails are shown correctly but most pictures only show a red screen. This is happening with my pictures taken with a Pixel 6a. On 1.6 all pictures could be viewed. 

**Screenshots**
![Screenshot_20240611-064044](https://github.com/leonlatsch/Photok/assets/48532778/daa9d9ff-70c8-49e1-a985-f7302f814930)
![Screenshot_20240611-064047](https://github.com/leonlatsch/Photok/assets/48532778/0519a118-c191-48e6-9ecd-39b4698f179d)

**Smartphone (please complete the following information):**
 - Device: Google Pixel 6a
 - OS: Android (CalyxOS 5.7.2)

**Comment by leonlatsch** [comment]
Are you using 1.7 or 1.7.1 ?

**Comment by prosch88** [comment]
1.7.0 from F-Droid
Tested 1.7.1 from Play Store with the same result.

**Comment by prosch88** [comment]
Just tested 1.7.1 from Github. Same result.

**Comment by leonlatsch** [comment]
Weird.
Would be nice if you could provide this image to me (if you are okay with it) so I can reproduce?

**Comment by leonlatsch** [comment]
Thank you for sending it to me :)

Unfortunately I was not able to reproduce this. I have made a fix. May you try it on your device?
[photok 1.7.1 - red screen fix.apk.zip](https://github.com/user-attachments/files/15784324/photok.1.7.1.-.red.screen.fix.apk.zip)

**Comment by prosch88** [comment]
Tested. Still not working.

**Comment by prosch88** [comment]
If I open the picture in the Gallery App and save a copy (edit / Export) the picture is viewable. The original one isn't.

**Comment by leonlatsch** [comment]
So actually the red color is the fallback and error for the component displaying the image. Are you able to provide logs?



> If I open the picture in the Gallery App and save a copy (edit / Export) the picture is viewable. The original one isn't.

Thats interesting.

**Comment by prosch88** [comment]
Is there a log export? My device isn't rooted.

**Comment by leonlatsch** [comment]
No unfortunately not. The only way currently is adb logcat

**Comment by prosch88** [comment]
I did send some logs. Also checked with version 1.6. The pictures are correctly shown ... unfortunately the 1.7 Backup isn't working with 1.6.

**Comment by leonlatsch** [comment]
Yes, backups are only compatible downwards. So newer backups are not compatible with old versions of Photok only the other way around

**Comment by leonlatsch** [comment]
> I did send some logs. Also checked with version 1.6. The pictures are correctly shown ... unfortunately the 1.7 Backup isn't working with 1.6.

Thank you, unfortunately they are very trimmed. Did you use the release version to create them? If yes, please try again with the apk I shared in this issue. It will show more logs. 
In the logs you sent, there is no error visible that could cause this

**Comment by leonlatsch** [comment]
I also received an email about this happening for other users. Marking as confirmed although I cannot reproduce right now

**Comment by leonlatsch** [comment]
@prosch88 

Does it only happen for images that are being taken with your camera app? Might have something to do with exif data

**Comment by prosch88** [comment]
Yes. This is happening for pictures take with the Pixel camera App.

**Comment by prosch88** [comment]
Some pictures (mostly small ones) are rendered. Most pictures with filesize 7MB and up are not...

**Comment by leonlatsch** [comment]
So I was not able to figure out anything new. I have created a new debug build for you that includes a special logger for the image loading library I use. If you could produce some logs again with this version, I think it will be very helpful :)

@prosch88 

[photok-debug-with-coil-logger.apk.zip](https://github.com/user-attachments/files/15842888/photok-debug-with-coil-logger.apk.zip)

**Comment by prosch88** [comment]
send you the log.

Opening the picture, this line is shown:

06-14 23:33:45.914 21752 21752 I RealImageLoader: 🚨 Failed - EncryptedImageRequestData(internalFileName=a6c0c112-b0fb-4612-b057-0d8871cdcc95.photok, mimeType=image/jpeg) - java.io.IOException: javax.crypto.AEADBadTagException: mac check in GCM failed

**Comment by KGriss** [comment]
I have the same issue with a storage backup I imported from 1.6.0 to 1.7.1. Most photos (not all) can't be rendered and they show a dark red screen.

**Comment by oa01** [comment]
I have the same issue on version 1.7.0
In my case every image gets rendered, except gif files

**Comment by leonlatsch** [comment]
> I have the same issue with a storage backup I imported from 1.6.0 to 1.7.1. Most photos (not all) can't be rendered and they show a dark red screen. 

Question to the others. Did you also import a backup before this happened? This would mean that restoring old backups is broken. 

@prosch88

**Comment by prosch88** [comment]
No. This happens also without importing a backup with fresh images loaded from Gallery.

**Comment by leonlatsch** [comment]
> No. This happens also without importing a backup with fresh images loaded from Gallery. 

Also when not importing new images?

Meaning images that were imported in 1.6.0 and have this error after updating to 1.7.0?

**Comment by leonlatsch** [comment]
> If I open the picture in the Gallery App and save a copy (edit / Export) the picture is viewable. The original one isn't. 

Also to others, would be nice to try this aswell. I still cannot reproduce this bug.

**Comment by UnrivaledScapegoat** [comment]
> > If I open the picture in the Gallery App and save a copy (edit / Export) the picture is viewable. The original one isn't.
> 
> Also to others, would be nice to try this aswell. I still cannot reproduce this bug.

I tried this with a gif. Edited the speed of the gif. Same issue. It does not render.

**Comment by leonlatsch** [comment]
News: I was finally able to reproduce this with the original image provided by @prosch88

**Comment by leonlatsch** [comment]
> 1.7.0 from F-Droid Tested 1.7.1 from Play Store with the same result.

Regarding this. I was not able to reproduce this with 1.7.0 On 1.7.1 its reproducable though

**Comment by leonlatsch** [comment]
Using the magic of git bisect I was able to Identify the bad commit as 

https://github.com/leonlatsch/Photok/commit/50771d845a7e023ed26f938950c039dca32f83af

Meaning in 1.7.0 this bug should not be happening and its related to using coil/compose for displaying the full size images. 
Most likely related to my implementation of the coil ImageFetcher that cannot handle large files encrypted with AES/GCM.
GCM encrypted files (such as Photoks) need to be read as a whole and cannot skip bytes. Maybe using a buffered source in the image fetcher causes this. Which would also explain the error `java.io.IOException: javax.crypto.AEADBadTagException: mac check in GCM failed`

TL;DR
Related to a change in 1.7.0. Will revert or try to fix it ASAP

**Comment by prosch88** [comment]
Just checked again with the F-Droid Version (1.7.0 is latest here) and it also shows this issue.

**Comment by leonlatsch** [comment]
I will provide you with a debug build that fixed the issue for me with your supplied image.

**Comment by leonlatsch** [comment]
> Just checked again with the F-Droid Version (1.7.0 is latest here) and it also shows this issue.

Yes. This might have been caused by a similar issue in Glide.

The core issue is that whenever the image was sampeld/buffered (bytes skipped). The GCM check in the decryption failed. So I implemented loading the whole image as a bitmap for now. Tell me if this fixes the issue for you :) For me it did

**Comment by prosch88** [comment]
Indeed this fixes the issue 👍

**Comment by leonlatsch** [comment]
Perfect. I will create a release tomorrow morning
